indexbs/indexbacked_bs.go

```diff
@@ -0,0 +1,233 @@
+package indexbs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/filecoin-project/dagstore"
+	blocks "github.com/ipfs/go-block-format"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/dagstore/shard"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+)
+
+var logbs = logging.Logger("dagstore-all-readblockstore")
+
+var ErrBlockNotFound = errors.New("block not found")
+
+var _ blockstore.Blockstore = (*IndexBackedBlockstore)(nil)
+
+// ErrNoShardSelected means that the shard selection function rejected all of the given shards.
+var ErrNoShardSelected = errors.New("no shard selected")
+
+// ShardSelectorF helps select a shard to fetch a cid from if the given cid is present in multiple shards.
+// It should return `ErrNoShardSelected` if none of the given shard is selected.
+type ShardSelectorF func(c cid.Cid, shards []shard.Key) (shard.Key, error)
+
+type accessorWithBlockstore struct {
+	sa *dagstore.ShardAccessor
+	bs dagstore.ReadBlockstore
+}
+
+// IndexBackedBlockstore is a read only blockstore over all cids across all shards in the dagstore.
+type IndexBackedBlockstore struct {
+	d            *dagstore.DAGStore
+	shardSelectF ShardSelectorF
+
+	bsStripedLocks  [256]sync.Mutex
+	blockstoreCache *lru.Cache // caches the blockstore for a given shard for shard read affinity i.e. further reads will likely be from the same shard. Maps (shard key -> blockstore).
+}
+
+func NewIndexBackedBlockstore(d *dagstore.DAGStore, shardSelector ShardSelectorF, maxCacheSize int) (blockstore.Blockstore, error) {
```

Comment:
More offline feedback on what remains before we can productionize this:

1. A control mechanism similar to Graphsync (Given a request from cidX for peerY -> should I run it) ?
2. Rate limiting similar to Graphsync.
3. Be able to run Bitswap retrieval as a separate process.